File: golang/q800/q867.go
package q800

func transpose(matrix [][]int) [][]int {
	if len(matrix) == 0 {
		return matrix
	}
	w := len(matrix)
	h := len(matrix[0])
	newMatrix := make([][]int, h)
	for i := 0; i < h; i++ {
		newMatrix[i] = make([]int, w)
	}
	for i := 0; i < w; i++ {
		for j := 0; j < h; j++ {
			newMatrix[j][i] = matrix[i][j]
		}
	}
	return newMatrix
}

File: hs/src/Q800/Q867.hs
module Q800.Q867
(
    transpose
) where

-- transpose :: [[a]] -> [[a]]
-- transpose [] = []
-- transpose [xs] = [[x] | x <- xs]
-- transpose (x:xs) = zipWith (:) x (transpose xs)

-- 特殊版本
transpose :: [[a]] -> [[a]]
transpose = foldr (zipWith (:)) (repeat [])
